src/query/src/dist_plan/predicate_extractor.rs

```diff
@@ -0,0 +1,1199 @@
+                // because if any part fails, the entire AND fails
+                let mut partition_constraints = Vec::new();
+                Self::extract_and_constraints(expr, partition_columns, &mut partition_constraints);
+
+                if partition_constraints.is_empty() {
+                    ExpressionCheckResult::Drop
+                } else if Self::expr_only_involves_partition_columns(expr, partition_columns) {
+                    // If the entire expression only involves partition columns, use as-is
+                    ExpressionCheckResult::UseAsIs(expr.clone())
+                } else {
+                    // Mixed expression: return only the partition parts
+                    ExpressionCheckResult::UsePartial(partition_constraints)
+                }
+            }
+            RestrictedOp::Or => {
+                // For OR expressions, we can only use them if ALL branches involve only partition columns
+                // because if any branch involves non-partition columns, we cannot safely prune
+                if Self::expr_only_involves_partition_columns(expr, partition_columns) {
+                    ExpressionCheckResult::UseAsIs(expr.clone())
+                } else {
+                    // Mixed OR expression: must drop entirely for safety
+                    ExpressionCheckResult::Drop
+                }
+            }
+            _ => {
+                // For comparison operations (=, <, >, etc.), check if they involve partition columns
+                if Self::operand_involves_partition_columns(expr.lhs(), partition_columns)
+                    || Self::operand_involves_partition_columns(expr.rhs(), partition_columns)
+                {
+                    if Self::expr_only_involves_partition_columns(expr, partition_columns) {
+                        ExpressionCheckResult::UseAsIs(expr.clone())
+                    } else {
+                        // This shouldn't happen for simple comparisons, but handle defensively
+                        ExpressionCheckResult::Drop
+                    }
+                } else {
+                    ExpressionCheckResult::Drop
+                }
+            }
+        }
+    }
+
+    /// Extract partition-related constraints from AND expressions recursively
+    fn extract_and_constraints(
+        expr: &PartitionExpr,
+        partition_columns: &HashSet<String>,
+        result: &mut Vec<PartitionExpr>,
+    ) {
+        if let RestrictedOp::And = expr.op() {
+            // Recursively process both sides of AND
+            Self::extract_constraints_from_operand(expr.lhs(), partition_columns, result);
+            Self::extract_constraints_from_operand(expr.rhs(), partition_columns, result);
+        } else {
+            // Non-AND expression: check if it involves partition columns
+            if Self::expr_involves_partition_columns(expr, partition_columns) {
+                result.push(expr.clone());
+            }
```

Comment:
Do we need to check if it only involves partition columns? What if this is an OR expr that involves both partition and non-partition columns?

Reply:
it will be dropped if the other side of OR is not partition column, see the second case "mixed_or_expression" of `test_mixed_partition_non_partition_expressions`